### PR TITLE
[SPARK-43530][PROTOBUF] Read descriptor file only once	

### PR DESCRIPTION
**Pull request by rangadi**
### What changes were proposed in this pull request?

Protobuf functions (`from_protobuf()` & `to_protobuf()`) take file path of a descriptor file and use that for constructing Protobuf descriptors. 
Main problem with how this is that the file is read many times (e.g. at each executor). This is unnecessary and error prone. E.g. file contents may be updated couple of days after a streaming query starts. That could lead to various errors. 

**The fix**: Use the byte content (which is serialized `FileDescritptorSet` proto). We read the content from the file once and carry the byte buffer. 

This also adds new API where we can pass the byte buffer directly. This is useful when the users fetch the content themselves and passes it to Protobuf functions. E.g. they could fetch it from S3, or extract it Python Protobuf classes. 

**Note to reviewers**: This includes a lot of updates to test files, mainly because the interface change to pass the buffer. I have left a few PR comments to help with the review.

### Why are the changes needed?
Described above.

### Does this PR introduce _any_ user-facing change?
Yes, this adds two new versions for `from_protobuf()` and `to_protobuf()` API that take Protobuf bytes rather than file path. 

### How was this patch tested?
 - Unit tests

### CODE REVIEW

**Comment by rangadi** [comment]
cc: @SandishKumarHN, @justaparth, @gengliangwang

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala [hunk 1]
```diff
@@ -26,14 +26,14 @@ import org.apache.spark.sql.types.{BinaryType, DataType}
 private[protobuf] case class CatalystDataToProtobuf(
     child: Expression,
     messageName: String,
-    descFilePath: Option[String] = None,
+    fileDescriptorSetBytesOpt: Option[Array[Byte]] = None,
```
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala:29]: File path is not carried anymore.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala [hunk 2]
```diff
@@ -58,14 +84,33 @@ object functions {
...
+   *   the binary column.
+   * @param messageName
+   *   the protobuf MessageName to look for in the descriptor set.
+   * @param fileDescriptorSetBytes
+   *   Serialized Protobuf descriptor (`FileDescriptorSet`). Typically contents of file created
+   *   using `protoc` with `--descriptor_set_out` and `--include_imports` options.
+   * @since 3.5.0
+   */
+  @Experimental
+  def from_protobuf(data: Column, messageName: String, fileDescriptorSetBytes: Array[Byte])
```
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:111]: This is an example new API that takes byte array as the argument. 
I will send a separate PR for Python API.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala [hunk 3]
```diff
@@ -223,28 +226,28 @@ private[sql] object ProtobufUtils extends Logging {
     }
   }
 
-  private def parseFileDescriptorSet(descFilePath: String): List[Descriptors.FileDescriptor] = {
-    var fileDescriptorSet: DescriptorProtos.FileDescriptorSet = null
+  def readDescriptorFileContent(filePath: String): Array[Byte] = {
```
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:229]: This util method is used in many places (including tests) to read the contents of the file.

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -47,11 +47,6 @@
     ],
     "sqlState" : "42846"
   },
-  "CANNOT_CONSTRUCT_PROTOBUF_DESCRIPTOR" : {
```
Comment (rangadi) [inline core/src/main/resources/error/error-classes.json:50]: This error was never required. It was masking another structured error.

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -125,7 +120,7 @@
   },
   "CANNOT_PARSE_PROTOBUF_DESCRIPTOR" : {
     "message" : [
-      "Error parsing file <descFilePath> descriptor byte[] into Descriptor object."
+      "Error parsing descriptor bytes into Protobuf FileDescriptorSet."
```
Comment (rangadi) [inline core/src/main/resources/error/error-classes.json:123]: We don't have the file name for this error. 
Failure to read from the file results in `QueryCompilationErrors.cannotFindDescriptorFileError()`

**Review by rangadi — COMMENTED** [review]
Added some notes for reviewers.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala [hunk 6]
```diff
@@ -148,8 +212,38 @@ object functions {
     messageName: String,
     descFilePath: String,
     options: java.util.Map[String, String]): Column = {
+    val fileContent = ProtobufUtils.readDescriptorFileContent(descFilePath)
```
Comment (SandishKumarHN) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:215]: Can't we change CatalystDataToProtobuf and ProtobufDataToCatalyst to accept descFilePath or fileDescriptorSetBytes as optional parameters instead of passing them in functions? Perhaps we could throw an error if both are not set?  This is just a thought.

Comment (SandishKumarHN) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:215]: nit
fileContent -> descriptorFileContent

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 1 above)
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala:29]: @advancedxy suggested using broadcast for distributing this buffer ([in the comment here](https://github.com/apache/spark/pull/41192#issuecomment-1551260315)): 

Moving that discussion to code comment since it has better threading.
@advancedxy most of the time size would be in kilobytes. It is possible some customers have huge protobuf repo and they bundled everything. 
Using broadcast is a good idea. I need to read up more on it. Can we do that as an implementation detail (so that users don't need be aware of it). We also need to make sure it is cleaned up.

I will add a comment. Do you think we can do that an enhancement later?
If we do it as implementation detail, it should work fine with Spark-Connect too.

**Comment by rangadi** [comment]
@advancedxy broadcast is an interesting idea. Lets continue the discussion in a code comment here: https://github.com/apache/spark/pull/41192#discussion_r1197264386

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 1 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala:29]: Sounds good. I will add comment. Out of curiosity: how does Spark know when a broadcast value is not needed in order to clean up?

**Comment by rangadi** [comment]
@gengliangwang PTAL when you get chance.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala (hunk 6 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:215]: Done.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 1 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala]: Added a TODO and created a ticket: https://issues.apache.org/jira/browse/SPARK-43578

**Comment by rangadi** [comment]
@SandishKumarHN Please approve if this looks good. Would like to merge this soon if there are no outstanding issues.

**Review by SandishKumarHN — APPROVED** [review]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala [hunk 7]
```diff
@@ -45,12 +53,36 @@ object functions {
       messageName: String,
       descFilePath: String,
       options: java.util.Map[String, String]): Column = {
+    val binaryFileDescSet = readDescriptorFileContent(descFilePath)
```
Comment (rangadi) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:56]: @LuciferYang I updated connect functions to use the buffer and change the protocol to send the buffer to the server rather the file name (as a result the file is only read on the client side). 
Updated the tests and added a ClientE2E test. PTAL.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:56]: Sorry for the late reply, will review this one later

Comment (rangadi) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:56]: @LuciferYang PTAL when you get a chance. I think the changes are fairly straight forward.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 8]
```diff
@@ -1452,29 +1452,29 @@ class SparkConnectPlanner(val session: SparkSession) {
...
-      val (descFilePathOpt, options) = if (argumentsCount == 2) {
+      val (binaryFileDescSetOpt, options) = if (argumentsCount == 2) {
         (None, Map.empty[String, String])
       } else if (argumentsCount == 3) {
         children(2) match {
-          case Literal(s, StringType) if s != null =>
-            (Some(s.toString), Map.empty[String, String])
+          case Literal(b, BinaryType) if b != null =>
+            (Some(b.asInstanceOf[Array[Byte]]), Map.empty[String, String])
           case UnresolvedFunction(Seq("map"), arguments, _, _, _) =>
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1468]: I  think we should refactor the error message due to the change from `descFilePath` to `binaryFileDescSetOpt` and the change in parameter type

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 9]
```diff
@@ -1452,29 +1452,29 @@ class SparkConnectPlanner(val session: SparkSession) {
...
             (None, ExprUtils.convertToMapData(CreateMap(arguments)))
           case other =>
             throw InvalidPlanInput(
               s"The valid type for the 3rd arg in $functionName is string or map, but got $other")
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1472]: ditto

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala [hunk 10]
```diff
@@ -195,7 +195,9 @@ class ProtoToParsedPlanTestSuite
   }
 
   private def removeMemoryAddress(expr: String): String = {
-    expr.replaceAll("@[0-9a-f]+,", ",")
+    expr
+      .replaceAll("@[0-9a-f]+,", ",")
+      .replaceAll("@[0-9a-f]+\\)", ")")
```
Comment (LuciferYang) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala:200]: What is the reason for this change?

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala [hunk 11]
```diff
@@ -26,14 +26,17 @@ import org.apache.spark.sql.types.{BinaryType, DataType}
 private[sql] case class CatalystDataToProtobuf(
     child: Expression,
     messageName: String,
-    descFilePath: Option[String] = None,
+    binaryFileDescriptorSet: Option[Array[Byte]] = None,
     options: Map[String, String] = Map.empty)
     extends UnaryExpression {
 
+  // TODO(SPARK-43578): binaryFileDescriptorSet could be very large in some cases. It is better
```
Comment (LuciferYang) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala:33]: Will `connect` also be changed like this TODO?

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala [hunk 12]
```diff
@@ -145,9 +147,10 @@ private[sql] object ProtobufUtils extends Logging {
    *  the `messageName` is treated as Java class name.
    * @return
    */
-  def buildDescriptor(messageName: String, descFilePathOpt: Option[String]): Descriptor = {
-    descFilePathOpt match {
-      case Some(filePath) => buildDescriptor(descFilePath = filePath, messageName)
+  def buildDescriptor(messageName: String, binaryFileDescriptorSet: Option[Array[Byte]])
```
Comment (LuciferYang) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:150]: Please fix the scaladoc of this function due to parameter changes

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala [hunk 13]
```diff
@@ -286,9 +289,9 @@ private[sql] object ProtobufUtils extends Logging {
   }
 
   /** Builds [[TypeRegistry]] with all the messages found in the descriptor file. */
-  private[protobuf] def buildTypeRegistry(descFilePath: String): TypeRegistry = {
+  private[protobuf] def buildTypeRegistry(descriptorBytes: Array[Byte]): TypeRegistry = {
```
Comment (LuciferYang) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:292]: Should the comment for this function also be corrected?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 14]
```diff
@@ -2251,10 +2252,8 @@ class PlanGenerationTestSuite
   //  1. cd connector/connect/common/src/main/protobuf/spark/connect
   //  2. protoc --include_imports --descriptor_set_out=../../../../test/resources/protobuf-tests/common.desc common.proto
   // scalastyle:on line.size.limit
-  private val testDescFilePath: String = java.nio.file.Paths
-    .get("../", "common", "src", "test", "resources", "protobuf-tests")
-    .resolve("common.desc")
-    .toString
+  private val testDescFilePath: String = s"${IntegrationTestUtils.sparkHome}/connector/" +
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:2255]: Maybe not in this one, but can we generate a `.desc` file to use and  remove dependencies on external file `common.desc`?

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 1 above)
Comment (LuciferYang) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala]: Is broadcast suitable for solving this problem? Will GC make it unusable?

**Comment by justaparth** [comment]
thanks for this change, definitely nice to read the descriptor once and pass it around to avoid drift if the descriptor file changes over time 🙏

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala (hunk 3 above)
Comment (pengzhon-db) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:229]: This function is duplicate of the one in connect client. Is there way to define them in a common place? I think this is a general question of connect. No need to address in this PR.

Comment (LuciferYang) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:229]: Actually, there is already a `common-utils` module in Spark and @amaliujia is working on it. I think connect client will also rely on the `common-utils` module In the future, then we can put these them in that module.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 8 above)
Comment (rangadi) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1468]: Good catch, updated the exception description.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 9 above)
Comment (rangadi) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1472]: Done.

**Review by rangadi — COMMENTED** [review]
@LuciferYang, @pengzhon-db thanks for the comments.
Addressed the comments and updated the PR. PTAL.

**Review by rangadi — COMMENTED** [review]
@LuciferYang, @pengzhon-db thanks for the comments.
Addressed the comments and updated the PR. PTAL.

**Review by rangadi — COMMENTED** [review]
@LuciferYang, @pengzhon-db thanks for the comments.
Addressed the comments and updated the PR. PTAL.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala [hunk 15]
```diff
@@ -199,4 +301,16 @@ object functions {
...
+
+  private def emptyOptions: java.util.Map[String, String] = Collections.emptyMap[String, String]()
+
+  private def readDescriptorFileContent(filePath: String): Array[Byte] = {
+    // This method is copied from org.apache.spark.sql.protobuf.util.ProtobufUtils
+    try {
+      FileUtils.readFileToByteArray(new File(filePath))
+    } catch {
+      case NonFatal(ex) =>
+        throw QueryCompilationErrors.cannotFindDescriptorFileError(filePath, ex)
```
[the remark this answers is not in the document]
Comment (pengzhon-db) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:313]: `readFileToByteArray` can throw a few different exceptions:
https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FileUtils.html#readFileToByteArray-java.io.File-

```
NullPointerException - if file is null.
FileNotFoundException - if the file does not exist, is a directory rather than a regular file, or for some other reason cannot be opened for reading.
IOException - if an I/O error occurs.
```
If it's IOException, it doesn't mean the descriptor file cannot be found, right? We can either throw a different error, or same error but with more generic error message/code

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 16]
```diff
@@ -2251,10 +2252,8 @@ class PlanGenerationTestSuite
...
+    "connect/common/src/test/resources/protobuf-tests/common.desc"
 
   test("from_protobuf messageClassName") {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:2258]: Run `build/mvn test -pl connector/connect/server ` ok? PlanGenerationTestSuite is ok , but the corresponding `ProtoToParsedPlanTestSuite` will fail.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 11 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala:33]: May be not. It doesn't save much, if any at all (we need to transfer the buffer to the server). 
We can treat those two separately even if we want handle on connect side.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala (hunk 12 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:151]: Updated.

**Review by rangadi — COMMENTED** [review]
@LuciferYang, @pengzhon-db addressed all remaining comments, I had missed a couple. PTAL. I would like to merge this soon.

**Review by rangadi — COMMENTED** [review]
@LuciferYang, @pengzhon-db addressed all remaining comments, I had missed a couple. PTAL. I would like to merge this soon.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:2258]: Thanks ~

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala (hunk 1 above)
Comment (rangadi) [reply connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/CatalystDataToProtobuf.scala]: I am resolving this for now. We can address in future PR for  https://issues.apache.org/jira/browse/SPARK-43578

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/protobuf/functions.scala:314]: The Javadoc there is incomplete. In github CI tests, this method throws `java.nio.file.NoSuchFileException`. Updated the PR.

```
sbt.ForkMain$ForkError: java.nio.file.NoSuchFileException: /non/existent/path.desc
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.Files.readAllBytes(Files.java:3152)
	at org.apache.commons.io.FileUtils.readFileToByteArray(FileUtils.java:2581)
	at org.apache.spark.sql.protobuf.utils.ProtobufUtils$.readDescriptorFileContent(ProtobufUtils.scala:232)
```

**Comment by rangadi** [comment]
@LuciferYang do you need any more changes for this? I fixed a test that fails only github CI. 
I would like to get this merged asap.

**Comment by rangadi** [comment]
@LuciferYang please let me know if you need any more changes. If none, I can ask someone else to take a look and merge as well.

**Comment by LuciferYang** [comment]
@rangadi 

image

**Comment by LuciferYang** [comment]
Let me take a look again

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 17]
```diff
@@ -1649,15 +1650,17 @@ class SparkConnectPlanner(val session: SparkSession) {
       // Protobuf-specific functions
       case "from_protobuf" if Seq(2, 3, 4).contains(fun.getArgumentsCount) =>
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:1651]: @rangadi after this change, Is there still a scene `fun.getArgumentsCount == 3`?

**Comment by LuciferYang** [comment]
@rangadi Please fix the compilation error first. For the rest, I only have this one question: https://github.com/apache/spark/pull/41192#discussion_r1206557463

**Review by rangadi — COMMENTED** [review]
@LuciferYang fixed the build error (it was a last minute change that broke the build)l
Case of 3 args still exists. PTAL.

**Review by pengzhon-db — APPROVED** [review]

**Comment by rangadi** [comment]
@LuciferYang all the test pass could you merge? I have multiple PRs waiting for this.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
Merged to master. Thanks @rangadi
